File: leetcode/62.不同路径/62.不同路径.go
package main

func uniquePaths(m int, n int) int {
	dp := make([][]int, m)
	// base case
	// 第一列 只有向下一种走法,初始化为 1
	for i := 0; i < m; i++ {
		dp[i] = make([]int, n)
		dp[i][0] = 1
	}
	// 第一行 只有向右一种走法,初始化为 1
	for j := 0; j < n; j++ {
		dp[0][j] = 1
	}
	// 动态转移方程
	// 达到i,j的路径数 = i-1,j(上一个格子) + i,j-1(左一个格子)
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			dp[i][j] = dp[i-1][j] + dp[i][j-1]
		}
	}
	return dp[m-1][n-1]
}

func main() {
	println(uniquePaths(3, 7))
}

File: hot/leetcode/62.不同路径/62.不同路径.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        """
        m是行，n是列
        动态规划
        base case:
            第一排or第一列的网格只能向右或向下，因此路径数都是1
            第一列 只有向下一种走法,初始化为 1
            第一行 只有向右一种走法,初始化为 1
        dp数组:
           dp[i][j] 到达i,j坐标的路径和
        动态转移方程:
           dp[i][j] = dp[i-1][j]+dp[i][j-1]
           每次移动向右或向下，因此每格移动路径数等于i-1,j和i,j-1的路径数之和
        """
        dp = [[0 for _ in range(n)] for _ in range(m)]
        for i in range(m):
            for j in range(n):
                if i == 0 or j == 0:
                    dp[i][j] = 1
                else:
                    dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
        return dp[-1][-1]


if __name__ == '__main__':
    print(Solution().uniquePaths(3, 7))
